File: sdks/go/gbz185/identity_code.go
package gbz185

import (
	"fmt"
	"regexp"
	"strings"
)

const AgentIdentityCodeOID = "1.2.156.3088"
const CurrentIdentityCodeVersion = "1"

var identityCodeNodeRE = regexp.MustCompile(`^[0-9A-Z]+$`)

type AgentIdentityCodeParts struct {
	OID                         string
	Version                     string
	RegistrationServiceProvider string
	RegistrationRequester       string
	OntologySerial              string
	InstanceSerial              string
}

func validateNode(value string, name string, maxLength int) error {
	if value == "" || len(value) > maxLength || !identityCodeNodeRE.MatchString(strings.ToUpper(value)) {
		return fmt.Errorf("%s must be 1-%d base36 characters", name, maxLength)
	}
	return nil
}

func ValidateIdentityCodeParts(parts AgentIdentityCodeParts) error {
	if parts.OID == "" {
		parts.OID = AgentIdentityCodeOID
	}
	if parts.Version == "" {
		parts.Version = CurrentIdentityCodeVersion
	}
	if parts.OID != AgentIdentityCodeOID {
		return fmt.Errorf("agent identity code OID must be %s", AgentIdentityCodeOID)
	}
	if parts.Version != CurrentIdentityCodeVersion {
		return fmt.Errorf("agent identity code version must be %s", CurrentIdentityCodeVersion)
	}
	if err := validateNode(parts.RegistrationServiceProvider, "registrationServiceProvider", 6); err != nil {
		return err
	}
	if err := validateNode(parts.RegistrationRequester, "registrationRequester", 6); err != nil {
		return err
	}
	if err := validateNode(parts.OntologySerial, "ontologySerial", 9); err != nil {
		return err
	}
	return validateNode(parts.InstanceSerial, "instanceSerial", 9)
}

func FormatIdentityCode(parts AgentIdentityCodeParts) (string, error) {
	if parts.OID == "" {
		parts.OID = AgentIdentityCodeOID
	}
	if parts.Version == "" {
		parts.Version = CurrentIdentityCodeVersion
	}
	parts.RegistrationServiceProvider = strings.ToUpper(parts.RegistrationServiceProvider)
	parts.RegistrationRequester = strings.ToUpper(parts.RegistrationRequester)
	parts.OntologySerial = strings.ToUpper(parts.OntologySerial)
	parts.InstanceSerial = strings.ToUpper(parts.InstanceSerial)

	if err := ValidateIdentityCodeParts(parts); err != nil {
		return "", err
	}
	return strings.Join([]string{
		parts.OID,
		parts.Version,
		parts.RegistrationServiceProvider,
		parts.RegistrationRequester,
		parts.OntologySerial,
		parts.InstanceSerial,
	}, "."), nil
}

func ParseIdentityCode(code string) (AgentIdentityCodeParts, error) {
	segments := strings.Split(code, ".")
	if len(segments) != 9 {
		return AgentIdentityCodeParts{}, fmt.Errorf("agent identity code must contain 9 dot-separated segments")
	}
	oid := strings.Join(segments[:4], ".")
	if oid != AgentIdentityCodeOID {
		return AgentIdentityCodeParts{}, fmt.Errorf("unsupported agent identity code OID: %s", oid)
	}
	version := segments[4]
	if version != CurrentIdentityCodeVersion {
		return AgentIdentityCodeParts{}, fmt.Errorf("unsupported agent identity code version: %s", version)
	}
	parts := AgentIdentityCodeParts{
		OID:                         AgentIdentityCodeOID,
		Version:                     CurrentIdentityCodeVersion,
		RegistrationServiceProvider: strings.ToUpper(segments[5]),
		RegistrationRequester:       strings.ToUpper(segments[6]),
		OntologySerial:              strings.ToUpper(segments[7]),
		InstanceSerial:              strings.ToUpper(segments[8]),
	}
	if err := ValidateIdentityCodeParts(parts); err != nil {
		return AgentIdentityCodeParts{}, err
	}
	return parts, nil
}

func ValidateIdentityCode(code string) bool {
	_, err := ParseIdentityCode(code)
	return err == nil
}

File: sdks/rust/src/identity_code.rs
use crate::Result;

pub const AGENT_IDENTITY_CODE_OID: &str = "1.2.156.3088";
pub const CURRENT_IDENTITY_CODE_VERSION: &str = "1";

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct AgentIdentityCodeParts {
    pub oid: String,
    pub version: String,
    pub registration_service_provider: String,
    pub registration_requester: String,
    pub ontology_serial: String,
    pub instance_serial: String,
}

impl AgentIdentityCodeParts {
    pub fn new(provider: &str, requester: &str, ontology: &str, instance: &str) -> Self {
        Self {
            oid: AGENT_IDENTITY_CODE_OID.to_string(),
            version: CURRENT_IDENTITY_CODE_VERSION.to_string(),
            registration_service_provider: provider.to_string(),
            registration_requester: requester.to_string(),
            ontology_serial: ontology.to_string(),
            instance_serial: instance.to_string(),
        }
    }
}

fn validate_node(value: &str, name: &str, max_length: usize) -> Result<()> {
    let valid = !value.is_empty()
        && value.len() <= max_length
        && value.chars().all(|ch| ch.is_ascii_alphanumeric());
    if !valid {
        return Err(format!("{name} must be 1-{max_length} base36 characters").into());
    }
    Ok(())
}

pub fn validate_identity_code_parts(parts: &AgentIdentityCodeParts) -> Result<()> {
    if parts.oid != AGENT_IDENTITY_CODE_OID {
        return Err(format!("agent identity code OID must be {AGENT_IDENTITY_CODE_OID}").into());
    }
    if parts.version != CURRENT_IDENTITY_CODE_VERSION {
        return Err(
            format!("agent identity code version must be {CURRENT_IDENTITY_CODE_VERSION}").into(),
        );
    }
    validate_node(
        &parts.registration_service_provider,
        "registrationServiceProvider",
        6,
    )?;
    validate_node(&parts.registration_requester, "registrationRequester", 6)?;
    validate_node(&parts.ontology_serial, "ontologySerial", 9)?;
    validate_node(&parts.instance_serial, "instanceSerial", 9)
}

pub fn format_identity_code(parts: &AgentIdentityCodeParts) -> Result<String> {
    let normalized = AgentIdentityCodeParts {
        oid: if parts.oid.is_empty() {
            AGENT_IDENTITY_CODE_OID.to_string()
        } else {
            parts.oid.clone()
        },
        version: if parts.version.is_empty() {
            CURRENT_IDENTITY_CODE_VERSION.to_string()
        } else {
            parts.version.clone()
        },
        registration_service_provider: parts.registration_service_provider.to_ascii_uppercase(),
        registration_requester: parts.registration_requester.to_ascii_uppercase(),
        ontology_serial: parts.ontology_serial.to_ascii_uppercase(),
        instance_serial: parts.instance_serial.to_ascii_uppercase(),
    };
    validate_identity_code_parts(&normalized)?;
    Ok([
        normalized.oid,
        normalized.version,
        normalized.registration_service_provider,
        normalized.registration_requester,
        normalized.ontology_serial,
        normalized.instance_serial,
    ]
    .join("."))
}

pub fn parse_identity_code(code: &str) -> Result<AgentIdentityCodeParts> {
    let segments: Vec<&str> = code.split('.').collect();
    if segments.len() != 9 {
        return Err("agent identity code must contain 9 dot-separated segments".into());
    }
    let oid = segments[0..4].join(".");
    if oid != AGENT_IDENTITY_CODE_OID {
        return Err(format!("unsupported agent identity code OID: {oid}").into());
    }
    let version = segments[4];
    if version != CURRENT_IDENTITY_CODE_VERSION {
        return Err(format!("unsupported agent identity code version: {version}").into());
    }
    let parts = AgentIdentityCodeParts {
        oid: AGENT_IDENTITY_CODE_OID.to_string(),
        version: CURRENT_IDENTITY_CODE_VERSION.to_string(),
        registration_service_provider: segments[5].to_ascii_uppercase(),
        registration_requester: segments[6].to_ascii_uppercase(),
        ontology_serial: segments[7].to_ascii_uppercase(),
        instance_serial: segments[8].to_ascii_uppercase(),
    };
    validate_identity_code_parts(&parts)?;
    Ok(parts)
}

pub fn validate_identity_code(code: &str) -> bool {
    parse_identity_code(code).is_ok()
}
